scripts/bench.rb

```diff
@@ -0,0 +1,310 @@
+      '--encrypt-keystore', 'false',
+      '--import-snapshot', '%<s>s',
+      '--halt-after-import'
+    ],
+    validate_command: [
+      './target/release/forest',
+      '--config', '%<c>s',
+      '--target-peer-count', '50',
+      '--encrypt-keystore', 'false',
+      '--import-snapshot', '%<s>s',
+      '--halt-after-import',
+      '--skip-load',
+      '--height', '%<h>s'
+    ]
+  }
+].freeze
+
+BENCH_PATHS = { tmpdir: Dir.mktmpdir('forest-benchs-') }.freeze
+
+# Provides human readable formatting to Numeric class
+class Numeric
+  def to_bibyte
+    syscall('numfmt', '--to=iec-i', '--suffix=B', '--format=%.2f', to_s)
+  end
+end
+
+def tmp_dir
+  BENCH_PATHS[:tmpdir]
+end
+
+def syscall(*command)
+  stdout, _stderr, status = Open3.capture3(*command)
+  return stdout.chomp if status.success?
+
+  raise "#{command}, #{status}"
+end
+
+def forest_version
+  syscall('./target/release/forest', '--version')
+end
+
+def default_config
+  toml_str = syscall('./target/release/forest-cli', 'config', 'dump')
+
+  default = TomlRB.parse(toml_str)
+  default['client']['data_dir'] = tmp_dir
+  default
+end
+
+def snapshot_dir
+  syscall('./target/release/forest-cli', 'snapshot', 'dir')
+end
+
+def db_dir
+  config = default_config
+  data_dir = config.dig('client', 'data_dir')
+  db = config.key?('rocks_db') ? 'rocksdb' : 'paritydb'
+
+  "#{data_dir}/mainnet/#{db}"
+end
+
+def db_size
+  size = syscall('du', '-h', db_dir)
+  size.split[0]
+end
+
+def hr(seconds)
+  seconds = seconds < MINUTE ? seconds.ceil(1) : seconds.ceil(0)
+  time = Time.at(seconds)
+  durfmt = "#{seconds > HOUR ? '%-Hh ' : ''}#{seconds < MINUTE ? '' : '%-Mm '}%-S#{seconds < MINUTE ? '.%1L' : ''}s"
+  time.strftime(durfmt)
+end
+
+def sample_proc(pid, metrics)
+  output = syscall('ps', '-o', 'rss,vsz', pid.to_s)
+  output = output.split("\n").last.split
+  metrics[:rss].push(output[0].to_i)
+  metrics[:vsz].push(output[1].to_i)
+end
+
+def proc_monitor(pid)
+  metrics = { rss: [], vsz: [] }
+  handle = Thread.new do
+    loop do
+      sample_proc(pid, metrics)
+      sleep 0.5
+    rescue StandardError => _e
+      break
+    end
+  end
+  [handle, metrics]
+end
+
+def exec_command_aux(command, metrics)
+  Open3.popen2(*command) do |i, o, t|
+    i.close
+
+    handle, proc_metrics = proc_monitor(t.pid)
+    o.each_line do |l|
+      print l
+    end
+
+    handle.join
+    metrics.merge!(proc_metrics)
+  end
+end
+
+def exec_command(command, dry_run)
+  puts "$ #{command.join(' ')}"
+  return {} if dry_run
+
+  metrics = {}
+  t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+  exec_command_aux(command, metrics)
+  t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+  metrics[:elapsed] = hr(t1 - t0)
+  metrics
+end
+
+def config_path(bench)
+  "#{tmp_dir}/#{bench[:name]}.toml"
+end
+
+def build_config_file(bench)
+  default = default_config
+  bench_config = bench[:config]
+  # TODO: Find a better way to merge (conserve the default keys)
```

Comment:
What do you mean?

Reply:
let's say I have:

```
{ key1: "value1", key2: "value2" }
```
And want to merge inside my bench config KVs, then at the end I will lose original ones and have:

```
{ key1: "bench_value1" }
```
It's correct because the config keys are the default ones, but I would prefer to have this result:
```
{ key1: "bench_value1", key2: "value2" }
```

Reply:
If I understand correctly, you want an immutable `merge`. In general, in Ruby, methods ending with `!` mutate the underlying collection, whereas the ones without are returning a new one. So, is just `merge` what you are looking for?

Reply:
Thanks for the explanation. But actually I want some `deep_merge` function but could not find one in the baselib 😞

Reply:
Will do!